### PR TITLE
Refactor nidcpower system tests

### LINKED ISSUE #1614: Refactor nidcpower system tests after adding independent channels support
With the addition of independent channels support in nidcpower, we have multiple PRs that need to update nidcpower system tests. This issue is to track a work item to update system tests once all PRs related to independent channels operation are merged into main.

#1594

### PR DESCRIPTION
**Pull request by smooresni**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
- [x] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.
- [x] I've added tests applicable for this pull request

### What does this Pull Request accomplish?

Refactors the nidcpower system test to accomplish the following goals:
1. Reduce the number of fixtures, thus making it easier to decide which fixtures future devs should use.
2. Enable a developer to mark at test to run on a legacy session, independent channels session, or both

This PR also adds additional tests focused on channel based repeated capabilities for both session types.

### List issues fixed by this Pull Request below, if any.

- Resolves #1614 

### What testing has been done?

Tests are passing on local machine.

### CODE REVIEW

**Comment by sbethur** [comment]
ok to test

**Comment by sbethur** [comment]
@smooresni, There are these warnings in the system test log:
```
============================== warnings summary ===============================
..\..\src\nidcpower\system_tests\test_system_nidcpower.py:79
  c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:79: PytestUnknownMarkWarning: Unknown pytest.mark.independent_channels_session_only - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.independent_channels_session_only

..\..\src\nidcpower\system_tests\test_system_nidcpower.py:85
  c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:85: PytestUnknownMarkWarning: Unknown pytest.mark.legacy_session_only - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.legacy_session_only

..\..\src\nidcpower\system_tests\test_system_nidcpower.py:91
  c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:91: PytestUnknownMarkWarning: Unknown pytest.mark.independent_channels_session_only - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.independent_channels_session_only

..\..\src\nidcpower\system_tests\test_system_nidcpower.py:98
  c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:98: PytestUnknownMarkWarning: Unknown pytest.mark.legacy_session_only - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.legacy_session_only

..\..\src\nidcpower\system_tests\test_system_nidcpower.py:181
  c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:181: PytestUnknownMarkWarning: Unknown pytest.mark.channels - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.channels('0')

..\..\src\nidcpower\system_tests\test_system_nidcpower.py:193
  c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:193: PytestUnknownMarkWarning: Unknown pytest.mark.channels - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.channels('0')
...
...
```

File: src/nidcpower/system_tests/test_system_nidcpower.py [hunk 1]
```diff
@@ -214,134 +245,149 @@ def test_measure_multiple(session):
...
     properties_used = ['unlock', 'voltage_level']
     sequence_name = 'my_sequence'
-    single_channel_session.source_mode = nidcpower.SourceMode.SEQUENCE
+    session.source_mode = nidcpower.SourceMode.SEQUENCE
     try:
-        single_channel_session.create_advanced_sequence(sequence_name=sequence_name, property_names=properties_used, set_as_active_sequence=True)
+        session.create_advanced_sequence(sequence_name=sequence_name, property_names=properties_used, set_as_active_sequence=True)
         assert False
     except TypeError:
         pass
```
Comment (sbethur) [inline src/nidcpower/system_tests/test_system_nidcpower.py:386]: (Ignore the the line I'm writing this comment for)

For tests that result in the DeprecationWarning, please check if there's a way to suppress the warning from reaching the system test log.

File: src/nidcpower/system_tests/test_system_nidcpower.py [hunk 2]
```diff
@@ -435,7 +484,9 @@ def test_init_backwards_compatibility_with_initialize_with_channels(resource_nam
         ('Dev1/0', ''),
         ('Dev1/0,Dev1/1', None),
         ('Dev1/0,Dev2/1', None),
-        ('Dev1/0,Dev2/1', '')
+        ('Dev1/0,Dev2/1', ''),
+        (['Dev1/0', 'Dev1/1'], ''),  # construct with list
```
Comment (sbethur) [inline src/nidcpower/system_tests/test_system_nidcpower.py:488]: Add a case where resource name is something like `Dev1/0-3`

File: src/nidcpower/system_tests/test_system_nidcpower.py [hunk 3]
```diff
@@ -474,3 +525,31 @@ def test_init_raises_driver_errors_for_invalid_arguments(resource_name, channels
...
+
+
+@pytest.mark.legacy_session_only
+def test_error_channel_name_not_allowed(session):
+    with pytest.raises(nidcpower.Error) as e:
+        session.channels['0'].instrument_model
+    assert e.value.code == -1074134971  # IVI_ERROR_CHANNEL_NAME_NOT_ALLOWED
+    assert e.value.description.find('The channel or repeated capability name is not allowed.') != -1
+
+
```
Comment (sbethur) [inline src/nidcpower/system_tests/test_system_nidcpower.py:553]: Some other test cases we should have:
 - `instruments` rep cap
 - A test similar to `test_repeated_capabilities_with_initiate`, but with multi-instrument session
- A test similar to `test_repeated_capabilities_with_initiate`, but with FQCN as resource_name
 - A test similar to `test_repeated_capabilities_with_initiate`, but with channels rep cap equal to output of `get_channel_names()`
 - Session opened on `Dev1/0` and method called with channels rep cap equal to `1`

Also, verify all API methods that were updated to add the Channels rep cap are tested at least once.

File: src/nidcpower/system_tests/test_system_nidcpower.py [hunk 4]
```diff
@@ -474,3 +525,31 @@ def test_init_raises_driver_errors_for_invalid_arguments(resource_name, channels
             # sessions, so we make a driver call on channels and ensure that errors
             session.channels[channels].output_function = nidcpower.OutputFunction.DC_VOLTAGE
     assert e.value.code == expected_error_code
```
Comment (sbethur) [inline src/nidcpower/system_tests/test_system_nidcpower.py:527]: Add this case to `test_init_raises_driver_errors_for_invalid_arguments`:
`'Dev1/0', None, False`

(Didn't see this case tested anywhere)

**Review by sbethur — CHANGES_REQUESTED** [review]

**Comment by smooresni** [comment]
> @smooresni, There are these warnings in the system test log:
> 
> ```
> ============================== warnings summary ===============================
> ..\..\src\nidcpower\system_tests\test_system_nidcpower.py:79
>   c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:79: PytestUnknownMarkWarning: Unknown pytest.mark.independent_channels_session_only - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
>     @pytest.mark.independent_channels_session_only
> 
> ..\..\src\nidcpower\system_tests\test_system_nidcpower.py:85
>   c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:85: PytestUnknownMarkWarning: Unknown pytest.mark.legacy_session_only - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
>     @pytest.mark.legacy_session_only
> 
> ..\..\src\nidcpower\system_tests\test_system_nidcpower.py:91
>   c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:91: PytestUnknownMarkWarning: Unknown pytest.mark.independent_channels_session_only - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
>     @pytest.mark.independent_channels_session_only
> 
> ..\..\src\nidcpower\system_tests\test_system_nidcpower.py:98
>   c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:98: PytestUnknownMarkWarning: Unknown pytest.mark.legacy_session_only - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
>     @pytest.mark.legacy_session_only
> 
> ..\..\src\nidcpower\system_tests\test_system_nidcpower.py:181
>   c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:181: PytestUnknownMarkWarning: Unknown pytest.mark.channels - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
>     @pytest.mark.channels('0')
> 
> ..\..\src\nidcpower\system_tests\test_system_nidcpower.py:193
>   c:\JenkinsAgent\6\workspace\nimi-bot\win32\nidcpower\src\nidcpower\system_tests\test_system_nidcpower.py:193: PytestUnknownMarkWarning: Unknown pytest.mark.channels - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
>     @pytest.mark.channels('0')
> ...
> ...
> ```

Registered the markers in `tox.ini` in 131234777f4958b64c79b5a9064810d4fcb9fa15

File: src/nidcpower/system_tests/test_system_nidcpower.py [hunk 5]
```diff
@@ -5,43 +5,66 @@
...
-@pytest.fixture(scope='function')
-def multi_instrument_session():
-    with nidcpower.Session(
-        '4162_01,4162_02', '', False, 'Simulate=1, DriverSetup=Model:4162; BoardType:PXIe', True
-    ) as simulated_session:
-        yield simulated_session
+def pytest_generate_tests(metafunc):
+    '''Parametrizes the "session" fixture by examining the the markers set for a test.
 
+    By default, the session fixture is parametrized so each test runs twice. Once with a legacy
```
[the remark this answers is not in the document]
Comment (smooresni) [reply src/nidcpower/system_tests/test_system_nidcpower.py]: ee9f9efb7dfc8455abafa247c902b5727eab0fec

File: src/nidcpower/system_tests/test_system_nidcpower.py [hunk 6]
```diff
@@ -414,7 +460,10 @@ def test_init_issues_deprecation_warnings(resource_name, channels, independent_c
         ('Dev1', None),
         ('Dev1', ''),
         ('Dev1', '0'),
-        ('Dev1', '0,1')
+        ('Dev1', '0,1'),
+        (['Dev1'], [0, 1]),
+        (('Dev1',), (0, 1)),
+        ('Dev1', range(2))
     ]
```
[the remark this answers is not in the document]
Comment (smooresni) [reply src/nidcpower/system_tests/test_system_nidcpower.py:506]: 89647306b07d8dbf6b5dba518c85a19f267a9d6c

File: src/nidcpower/system_tests/test_system_nidcpower.py (hunk 1 above)
Comment (smooresni) [reply src/nidcpower/system_tests/test_system_nidcpower.py:414]: 79c92402b3006af63f64e5433a682542bbb13e31

File: src/nidcpower/system_tests/test_system_nidcpower.py (hunk 2 above)
Comment (smooresni) [reply src/nidcpower/system_tests/test_system_nidcpower.py:527]: 0e2734304b014bb6c7392e0134fa804d77611466

File: tox.ini [hunk 7]
```diff
@@ -174,4 +174,11 @@ norecursedirs = system_tests examples .* build dist CVS _darcs {arch} *.egg venv
 junit_suite_name = nimi-python
 python_files = *.py
 junit_family = xunit1
-
+markers = # Defines custom markers used by nidcpower system tests. Prevents PytestUnknownMarkWarning.
```
Comment (sbethur) [inline tox.ini:177]: This tox.ini is not used for running system tests, does this actually get rid of the warning?!

This code-generated tox.ini is what is used for running system tests on nimi-bot agents: https://github.com/ni/nimi-python/blob/master/generated/nidcpower/tox-system_tests.ini

Comment (smooresni) [reply tox.ini]: I've moved the custom markers to the mako template in https://github.com/ni/nimi-python/pull/1624/commits/7b97e591e765b91e2a0e785448994a75aaf30398

File: src/nidcpower/system_tests/test_system_nidcpower.py (hunk 3 above)
Comment (smooresni) [reply src/nidcpower/system_tests/test_system_nidcpower.py:626]: Added coverage in cb25392bc1430e3e03e2d741363eebfea78a4403 and 7696d83dbb5fc0e6b79e694bac38807e0830f8a1

File: src/nidcpower/system_tests/test_system_nidcpower.py (hunk 4 above)
Comment (smooresni) [reply src/nidcpower/system_tests/test_system_nidcpower.py]: Added coverage in cb25392bc1430e3e03e2d741363eebfea78a4403 and 7696d83dbb5fc0e6b79e694bac38807e0830f8a1

**Comment by smooresni** [comment]
Currently have a test set to be skipped. We should wait to merge until #1626 is merged.

**Comment by marcoskirsch** [comment]
>Currently have a test set to be skipped. We should wait to merge until #1626 is merged.

It's merged

File: src/nidcpower/system_tests/test_system_nidcpower.py [hunk 8]
```diff
@@ -458,19 +487,211 @@ def test_init_raises_value_error_for_multi_instrument_resource_name_and_channels
...
+@pytest.mark.legacy_session_only
+def test_error_channel_name_not_allowed_in_legacy_session(session):
+    with pytest.raises(nidcpower.Error) as e:
+        session.channels['0'].reset()
+    assert e.value.code == -1074118494  # NIDCPOWER_ERROR_CHANNEL_NAME_NOT_ALLOWED_IN_OBSOLETE_SESSION
+    assert e.value.description.find('The channel name string must represent all channels in the session because the session was not initialized with independent channels. To specify a subset of channels for this function, first initialize the session with independent channels.') != -1
+
+
+@pytest.mark.legacy_session_only
+def test_error_channel_name_not_allowed(session):
```
Comment (sbethur) [inline src/nidcpower/system_tests/test_system_nidcpower.py:550]: I think this test works with Independent Channels session too. Any particular reason for using `legacy_session_only`?

**Review by sbethur — APPROVED** [review]
Great job adding new tests and cleaning up existing tests!

File: src/nidcpower/system_tests/test_system_nidcpower.py (hunk 8 above)
Comment (smooresni) [reply src/nidcpower/system_tests/test_system_nidcpower.py:550]: For legacy sessions, `instrument_model` isn't expecting a channel string to be supplied.
For independent channels sessions, it expects the instrument name to be supplied.

So for an independent channels session, it throws a different error:
`"The specified device name cannot be found."`

Do you think we should also test for the independent channels error?

Comment (sbethur) [reply src/nidcpower/system_tests/test_system_nidcpower.py:550]: Ah, you're right. On the Independent Channels session, Python layer doesn't really differentiate between the various repeated capability containers and passes it down to the driver Runtime. So, we don't need to test it with  Independent Channels session.

**Review by marcoskirsch — APPROVED** [review]

**Comment by marcoskirsch** [comment]
ok to test

**Review by marcoskirsch — APPROVED** [review]

**Pull request merged**